### ISSUE TITLE
docker-compose - Read timed out

### ISSUE THREAD

**Issue by yurinnick** [kernelci/kernelci-pipeline#290]
When starting services locally, most `kernelci-pipeline` services fail with error:
```
kernelci-pipeline-notifier            |     httplib_response = conn.getresponse()
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/http/client.py", line 1374, in getresponse
kernelci-pipeline-notifier            |     response.begin()
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/http/client.py", line 318, in begin
kernelci-pipeline-notifier            |     version, status, reason = self._read_status()
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/http/client.py", line 279, in _read_status
kernelci-pipeline-notifier            |     line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/socket.py", line 705, in readinto
kernelci-pipeline-notifier            |     return self._sock.recv_into(b)
kernelci-pipeline-notifier            | TimeoutError: timed out
kernelci-pipeline-notifier            |
kernelci-pipeline-notifier            | During handling of the above exception, another exception occurred:
kernelci-pipeline-notifier            |
kernelci-pipeline-notifier            | Traceback (most recent call last):
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/site-packages/requests/adapters.py", line 486, in send
kernelci-pipeline-notifier            |     resp = conn.urlopen(
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 798, in urlopen
kernelci-pipeline-notifier            |     retries = retries.increment(
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/site-packages/urllib3/util/retry.py", line 550, in increment
kernelci-pipeline-notifier            |     raise six.reraise(type(error), error, _stacktrace)
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/site-packages/urllib3/packages/six.py", line 770, in reraise
kernelci-pipeline-notifier            |     raise value
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 714, in urlopen
kernelci-pipeline-notifier            |     httplib_response = self._make_request(
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 468, in _make_request
kernelci-pipeline-notifier            |     self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
kernelci-pipeline-notifier            |   File "/usr/local/lib/python3.10/site-packages/urllib3/connectionpool.py", line 357, in _raise_timeout
kernelci-pipeline-notifier            |     raise ReadTimeoutError(
kernelci-pipeline-notifier            | urllib3.exceptions.ReadTimeoutError: HTTPConnectionPool(host='172.17.0.1', port=8001): Read timed out. (read timeout=60.0)
```

API side looks fine though

```
kernelci-api          | INFO:     172.19.0.1:58296 - "POST /latest/token HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33140 - "POST /latest/subscribe/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33160 - "POST /latest/subscribe/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33164 - "POST /latest/subscribe/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33174 - "GET /latest/nodes?timeout__lt=2023-07-27T07%3A11%3A08.237166&state=running&limit=100&offset=0 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33178 - "GET /latest/nodes?timeout__lt=2023-07-27T07%3A11%3A08.237166&state=available&limit=100&offset=0 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33182 - "GET /latest/nodes?timeout__lt=2023-07-27T07%3A11%3A08.237166&state=closing&limit=100&offset=0 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33198 - "POST /latest/subscribe/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33208 - "GET /latest/nodes?state=closing&limit=100&offset=0 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33214 - "GET /latest/nodes?state=available&holdoff__lt=2023-07-27T07%3A11%3A08.716990&limit=100&offset=0 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33218 - "POST /latest/subscribe/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33236 - "POST /latest/subscribe/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33254 - "POST /latest/subscribe/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33286 - "GET /latest/count?revision.commit=0abdfc23b2986bbfa1dd813a28b8730d7aefb0f4 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33294 - "POST /latest/node HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33270 - "GET /latest/listen/61 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33252 - "GET /latest/listen/60 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33230 - "GET /latest/listen/59 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33206 - "GET /latest/listen/58 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33170 - "GET /latest/listen/57 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33166 - "GET /latest/listen/56 HTTP/1.1" 200 OK
kernelci-api          | INFO:     172.19.0.1:33150 - "GET /latest/listen/55 HTTP/1.1" 200 OK
```

**Comment by gctucker** [comment]
Ah, I see what happened there.  Can you try adding a keep-alive period in your environment file?  On the staging deployment, in `.env` we have:
```
KEEP_ALIVE_PERIOD=45
```

Alternatively, you can disable the timeout in the client side YAML config:
```yaml
api_configs:

  docker-host:
    url: http://172.17.0.1:8001
    timeout: 0
```

Initially there was no keep-alive feature and no timeout in the client library.  That worked fine with a local docker-compose setup, so I think just disabling it should work too.  If that's the case then we might change the default to `0` in the code as right now the default is 60 seconds.

Then when we deployed the staging API hosted on a VM in Azure, we started seeing timeouts while client services were waiting for pub/sub events.  This is a TCP query that stays open until an event is received.  To avoid hitting his kind of infrastructure timeout, we added the "keep-alive" feature which sends a `BEEP` message periodically (every 45s on staging) which then gets filtered out in the client library.  The default setting is still to not have keep-alive as it's not normally required and causes extra traffic.  However, when we upgraded some components including `requests` and `pylint` in on the client side (in `kernelci-core`), there was a new warning that doing requests without a timeout could lead to some issues if the server never replied.  That seemed like a far thing to address so we now have a `timeout` config setting in the client library.  As a result, the default is now to not have keep-alive on the API server side but a 60s timeout on the client side.  The safest way to solve this would probably be to enable 45s keep-alive in the API and leave the 60s timeout in the client library, but the optimal way to solve it would be to disable the timeout in the client library.  I think the side-effects of the keep-alive are really minor so we probably should enable it by default to avoid having other users tripping over the issue when deploying the API in their infrastructure.

But please confirm either or both of the suggested setting tweaks above resolves your issue locally to be sure.

**Comment by yurinnick** [comment]
Unfortunately none of this methods worked, adding parameter to `.env` didn't have any effect, and setting `timeout: 0` caused request to fail immediately. However, setting `timeout: 3600` did the trick.

**Comment by gctucker** [comment]
OK thanks.  It's a shame a timeout of 0 doesn't mean "no timeout".  Maybe `null` would have worked.  I'm pretty sure adding the keep-alive value on the API side should work though as that's what we have on the staging instance.  I'll take a closer look.

**Comment by gctucker** [comment]
Alright I've reproduced the issue locally and confirmed that setting a 45s keep-alive period on the API side solves the issue.  I've created a PR to set it as a default value.  It's still possible to override it with the environment variable, if users want to disable it.  I haven't looked into how we can disable the timeout on the client side though (best bet is `null` in YAML would translate to `None` in Python and that would do the trick but tbc).
